File: 0160-intersection-of-two-linked-lists.cpp
/*
160. Intersection of Two Linked Lists

Submitted: May 31, 2025

Runtime: 40 ms (beats 37.90%)
Memory: 20.66 MB (beats 17.77%)
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!headA || !headB) return nullptr;
        vector<ListNode*> listA, listB;
        for (auto p = headA; p; p = p->next) listA.push_back(p);
        for (auto p = headB; p; p = p->next) listB.push_back(p);
        if (listA.back() != listB.back()) return nullptr;
        if (listB.size() > listA.size()) swap(listA, listB);
        for (
            auto itA = listA.begin() + (listA.size() - listB.size()), itB = listB.begin();
            itA != listA.end(); ++itA, ++itB
        ) {
            if (*itA == *itB) return *itA;
        }
        throw new runtime_error("Error");
    }
};

File: 0160-intersection-of-two-linked-lists.py
"""
160. Intersection of Two Linked Lists

Submitted: May 31, 2025

Runtime: 84 ms (beats 62.92%)
Memory: 27.87 MB (beats 45.04%)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

def get_nodes(head):
    while head:
        yield head
        head = head.next

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        if not headA or not headB: return None
        listA = list(get_nodes(headA))
        listB = list(get_nodes(headB))
        if listA[-1] is not listB[-1]: return None
        if len(listA) < len(listB):
            listA, listB = listB, listA
        for (a, b) in zip(listA[len(listA) - len(listB):], listB):
            if a is b: return a